File: src/01_basics/basics_cpp/src/action_server.cpp
// TUTORIAL 3a — Actions: Server
//
// ros2 run basics_cpp action_server
// ros2 run basics_cpp action_client    (separate terminal)

#include <memory>
#include <thread>
#include <chrono>
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "example_interfaces/action/fibonacci.hpp"

using Fibonacci = example_interfaces::action::Fibonacci;
using GoalHandle = rclcpp_action::ServerGoalHandle<Fibonacci>;
using namespace std::chrono_literals;

class CountServer : public rclcpp::Node
{
public:
  CountServer() : Node("count_server")
  {
    // === CHECKPOINT: actions ===
    action_server_ = rclcpp_action::create_server<Fibonacci>(
      this, "count_up",
      [](const rclcpp_action::GoalUUID &, std::shared_ptr<const Fibonacci::Goal>) {
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
      },
      [](std::shared_ptr<GoalHandle>) {
        return rclcpp_action::CancelResponse::ACCEPT;
      },
      [this](std::shared_ptr<GoalHandle> gh) {
        std::thread([this, gh]() { execute(gh); }).detach();
      });
    RCLCPP_INFO(get_logger(), "Action server /count_up ready");
    // === END CHECKPOINT: actions ===
  }

private:
  // === CHECKPOINT: actions ===
  void execute(std::shared_ptr<GoalHandle> goal_handle)
  {
    int target = goal_handle->get_goal()->order;
    RCLCPP_INFO(get_logger(), "Goal received: count to %d", target);

    auto feedback = std::make_shared<Fibonacci::Feedback>();
    for (int i = 0; i <= target; ++i) {
      if (goal_handle->is_canceling()) {
        goal_handle->canceled(std::make_shared<Fibonacci::Result>());
        return;
      }
      feedback->sequence = {i};
      goal_handle->publish_feedback(feedback);
      RCLCPP_INFO(get_logger(), "Feedback: %d/%d", i, target);
      std::this_thread::sleep_for(1s);
    }

    auto result = std::make_shared<Fibonacci::Result>();
    result->sequence = {target};
    goal_handle->succeed(result);
    RCLCPP_INFO(get_logger(), "Goal succeeded — reached %d", target);
  }

  rclcpp_action::Server<Fibonacci>::SharedPtr action_server_;
  // === END CHECKPOINT: actions ===
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CountServer>());
  rclcpp::shutdown();
}

File: src/01_basics/basics_py/basics_py/action_server.py
"""
TUTORIAL 3a — Actions: Server

Counts from 0 to a target number, publishing feedback at each step.
Use this to see goal / feedback / result lifecycle.

    ros2 run basics action_server
    ros2 run basics action_client    (separate terminal)
"""

import time
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from example_interfaces.action import Fibonacci   # reuse builtin: goal=order, feedback=sequence, result=sequence


# We repurpose the Fibonacci action type for simplicity:
#   goal.order      → count target
#   feedback.sequence  → [current_count]
#   result.sequence    → [final_count]

class CountServer(Node):
    def __init__(self):
        super().__init__('count_server')

        # === CHECKPOINT: actions ===
        self._action_server = ActionServer(
            self,
            Fibonacci,
            'count_up',
            self.execute_callback,
        )
        self.get_logger().info('Action server /count_up ready')
        # === END CHECKPOINT: actions ===

    # === CHECKPOINT: actions ===
    def execute_callback(self, goal_handle):
        target = goal_handle.request.order
        self.get_logger().info(f'Goal received: count to {target}')

        feedback = Fibonacci.Feedback()
        for i in range(target + 1):
            feedback.sequence = [i]
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f'Feedback: {i}/{target}')
            time.sleep(1.0)

        goal_handle.succeed()
        result = Fibonacci.Result()
        result.sequence = [target]
        self.get_logger().info(f'Goal succeeded — reached {target}')
        return result
    # === END CHECKPOINT: actions ===


def main(args=None):
    rclpy.init(args=args)
    node = CountServer()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
